### PR TITLE
argocd: new params for UpdateSpec and CreateUpdate actions. fix timeout issues

### PR DESCRIPTION
**Pull request by hemant0308**
* Add options waitForSync,syncTimeout,watchHealth parameters for UpdateSpec and CreateUpdate operations
* Fixed timeout issues for waitForSync method by moving BuffereReader instance creation statement from try-with-resource block.

### CODE REVIEW

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java [hunk 1]
```diff
@@ -50,23 +51,33 @@ public CallRetry(Callable<R> mainAttempt, Callable<Optional<R>> fallbackAttempt)
...
 
         int attemptsMade = 0;
         while (attemptsMade++ < maxTries) {
             try {
-                return mainAttempt.call();
+                R out = mainAttempt.call();
+                log.info("after main attempt {}", new Date().toString());
+                return out;
             } catch (Exception e) {
+                if(exceptionsToNotRetry.stream().anyMatch(exceptionToNotRetry -> e.getClass().equals(exceptionToNotRetry))) {
```
Comment (benbroadaway) [inline tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java:64]: `Class.isInstance()` is preferable for handling possible instances of subclasses. e.g. we'd still want to skip retries if a subclass of `SocketException` is caught.
```java
if (exceptionsToNotRetry.stream().anyMatch(c -> c.isInstance(e))) {
    // ...
}
```

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java [hunk 2]
```diff
@@ -23,23 +23,24 @@
...
-import java.util.Collections;
-import java.util.Optional;
+import java.util.*;
 import java.util.concurrent.Callable;
 
 public class CallRetry<R> {
     private static final Logger log = LoggerFactory.getLogger(CallRetry.class);
     private final Callable<R> mainAttempt;
     private final Collection<Callable<Optional<R>>> fallbackAttempts;
+    private final List<Class<? extends Exception>> exceptionsToNotRetry;
```
Comment (benbroadaway) [inline tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java:33]: Prefer a `Set` so entries are not repeated.

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java [hunk 3]
```diff
@@ -50,23 +51,33 @@ public CallRetry(Callable<R> mainAttempt, Callable<Optional<R>> fallbackAttempt)
      * @return Results of main call, or fallback result if data is present after main failure
      */
     public R attemptWithRetry(int maxTries) {
+        log.info("attempt with retry start {}", new Date().toString());
```
Comment (benbroadaway) [inline tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java:54]: Any particular reason for all the `new Date().toString()` logging? The log line already has the date time.

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/TaskParamsImpl.java [hunk 4]
```diff
@@ -847,7 +884,7 @@ public String propagationPolicy() {
         }
     }
 
-    private static class SyncParamsImpl extends TaskParamsImpl implements SyncParams {
+    public static class SyncParamsImpl extends TaskParamsImpl implements SyncParams {
```
Comment (benbroadaway) [inline tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/TaskParamsImpl.java:887]: Is this change to `public` on purpose? It's better to keep these implementations private so if/when we need to refactor we can just do it without worrying about public dependents.

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java [hunk 5]
```diff
@@ -181,7 +181,9 @@ private TaskResult processUpdateSpecAction(TaskParams.UpdateSpecParams in) throw
             appSpec = ConfigurationUtils.deepMerge(appSpec, in.spec());
             V1alpha1ApplicationSpec specObject = objectMapper.mapToModel(appSpec, V1alpha1ApplicationSpec.class);
             V1alpha1ApplicationSpec result = client.updateAppSpec(in.app(), specObject);
-
+            if(in.waitForSync()) {
```
Comment (benbroadaway) [inline tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java:184]: nitpick: space between `if` and `(`

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java [hunk 6]
```diff
@@ -261,8 +263,10 @@ private TaskResult processCreateAction(TaskParams.CreateUpdateParams in) throws
             V1alpha1Application application = objectMapper.buildApplicationObject(in);
             ArgoCdClient client = new ArgoCdClient(in);
             V1alpha1Application app = client.createApp(application, in.upsert());
-            app = client.waitForSync(in.app(), app.getMetadata().getResourceVersion(), in.syncTimeout(),
-                    toWatchParams(false));
+            if(in.waitForSync()){
```
Comment (benbroadaway) [inline tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java:266]: nitpick: space between `if` and `(`

**Review by benbroadaway — CHANGES_REQUESTED** [review]

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java (hunk 6 above)
Comment (hemant0308) [reply tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java]: Updated

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java (hunk 5 above)
Comment (hemant0308) [reply tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/ArgoCdTask.java]: Updated

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java (hunk 3 above)
Comment (hemant0308) [reply tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java]: Removed all unnecessary logs.

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java (hunk 2 above)
Comment (hemant0308) [reply tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java]: Updated with set

File: tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java (hunk 1 above)
Comment (hemant0308) [reply tasks/argocd/src/main/java/com/walmartlabs/concord/plugins/argocd/CallRetry.java]: Updated

**Pull request merged**